tests/integration/new_relations/test_new_relations.py

```diff
@@ -0,0 +1,297 @@
+        ops_test,
+        APPLICATION_APP_NAME,
+        MULTIPLE_DATABASE_CLUSTERS_RELATION_NAME,
+        relation_id=first_cluster_relation.id,
+    )
+    another_application_connection_string = await build_connection_string(
+        ops_test,
+        APPLICATION_APP_NAME,
+        MULTIPLE_DATABASE_CLUSTERS_RELATION_NAME,
+        relation_id=second_cluster_relation.id,
+    )
+    assert application_connection_string != another_application_connection_string
+
+
+async def test_an_application_can_connect_to_multiple_aliased_database_clusters(
+    ops_test: OpsTest, database_charm
+):
+    """Test that an application can connect to different clusters of the same database."""
+    # Relate the application with both database clusters
+    # and wait for them exchanging some connection data.
+    await asyncio.gather(
+        ops_test.model.add_relation(
+            f"{APPLICATION_APP_NAME}:{ALIASED_MULTIPLE_DATABASE_CLUSTERS_RELATION_NAME}",
+            DATABASE_APP_NAME,
+        ),
+        ops_test.model.add_relation(
+            f"{APPLICATION_APP_NAME}:{ALIASED_MULTIPLE_DATABASE_CLUSTERS_RELATION_NAME}",
+            ANOTHER_DATABASE_APP_NAME,
+        ),
+    )
+    await ops_test.model.wait_for_idle(apps=APP_NAMES, status="active")
+
+    # Retrieve the connection string to both database clusters using the relation aliases
+    # and assert they are different.
+    application_connection_string = await build_connection_string(
+        ops_test,
+        APPLICATION_APP_NAME,
+        ALIASED_MULTIPLE_DATABASE_CLUSTERS_RELATION_NAME,
+        relation_alias="cluster1",
+    )
+    another_application_connection_string = await build_connection_string(
+        ops_test,
+        APPLICATION_APP_NAME,
+        ALIASED_MULTIPLE_DATABASE_CLUSTERS_RELATION_NAME,
+        relation_alias="cluster2",
+    )
+    assert application_connection_string != another_application_connection_string
+
+
+async def test_an_application_can_request_multiple_databases(ops_test: OpsTest, application_charm):
+    """Test that an application can request additional databases using the same interface."""
+    # Relate the charms using another relation and wait for them exchanging some connection data.
+    await ops_test.model.add_relation(
+        f"{APPLICATION_APP_NAME}:{SECOND_DATABASE_RELATION_NAME}", DATABASE_APP_NAME
+    )
+    await ops_test.model.wait_for_idle(apps=APP_NAMES, status="active")
+
+    # Get the connection strings to connect to both databases.
+    first_database_connection_string = await build_connection_string(
+        ops_test, APPLICATION_APP_NAME, FIRST_DATABASE_RELATION_NAME
+    )
+    second_database_connection_string = await build_connection_string(
+        ops_test, APPLICATION_APP_NAME, SECOND_DATABASE_RELATION_NAME
+    )
+
+    # Assert the two application have different relation (connection) data.
+    assert first_database_connection_string != second_database_connection_string
```

Comment:
might be worth adding a test that scales up/down PG and then verifies that the relation data is properly updated and can still perform CRUD ops etc. 

Specifically add 2 units, remove original units so that PG has none of the units it was originally deployed with. Then verify that application can still connect to PG.

^ this suggestion is not for the endpoint tests which you mention cannot do do this scaling due to patroni. But instead for ensuring that relation departed/joined events update the connection URI and other relation data. (Can be done in a future PR)

Reply:
Good suggestion, Mia. I implemented the new test (with fixes figured out when implementing it) in the [next PR](https://github.com/canonical/postgresql-operator/pull/17). It is still a draft because I am checking why scaling down works locally but not when the test is running on Github CI.

I also merged some missing changes from the previous PR and the tests of this PR as working now.